datalad_next/iter_collections/annexworktree.py

```diff
@@ -0,0 +1,194 @@
+      untracked empty directories. Any untracked content is yielded as
+      a ``PurePosixPath``.
+    link_target: bool, optional
+      This flag behaves analog to that of ``iter_gitworktree()``. In addition,
+      enabling link target reporting will provide the location
+      of an annex object (annexed file content) via the
+      ``AnnexWorktreeItem.annexobjpath`` property. Annex object path reporting
+      is supported whether or not a particular key is locally present.
+    fp: bool, optional
+      If ``True``, each file-type item includes a file-like object
+      to access the file's content. This file handle will be closed
+      automatically when the next item is yielded.
+
+    Yields
+    ------
+    :class:`AnnexWorktreeItem` or `AnnexWorktreeFileSystemItem`
+    """
+    # get a lookup mapping platform-paths to annex key and bytesize
+    # for any path connected to an annex key
+    annex = _get_annexinfo(path)
+
+    # we use git-annex-examinekey for annex object location reporting.
+    # we cannot use git-annex-contentlocation, because it only reports on
+    # present annex objects, and here we also need to report on would-be
+    # locations
+    with annexjson_batchcommand(
+            ['git', 'annex', 'examinekey', '--json', '--batch'],
+            cwd=path,
+    ) as exkey:
+        for item in iter_gitworktree(
+            path=path,
+            untracked=untracked,
+            # we relay the link_target flag to ensure homogeneous behavior
+            # with iter_gitworktree(), but will also do further processing
+            # below
+            link_target=False,
+            # we need to account for annexed files and would rather open
+            # their content in the annex, so we have to do this separately
+            fp=False
+        ):
+            ainfo = annex.get(item.name, {})
+            akey = ainfo.get('key')
+            aobjpath = None
+            if link_target and akey:
+                keyprops = exkey(bytes(akey, encoding="utf-8") + b'\n')
+                aobjpath = keyprops.get('objectpath')
+                if aobjpath:
+                    aobjpath = PurePath(aobjpath)
+
+            asize = ainfo.get('bytesize')
+            if asize is not None:
+                asize = int(asize)
+            yield AnnexWorktreeItem(
+                name=item.name,
+                gitsha=item.gitsha,
+                gittype=item.gittype,
+                annexkey=ainfo.get('key'),
+                annexsize=asize,
+                annexobjpath=aobjpath,
+            )
+
+
+def _get_annexinfo(path: Path) -> dict:
+    from datalad.support.annexrepo import GeneratorAnnexJsonNoStderrProtocol
+
+    map = {
+        'bytesize': lambda x: x,
+        'key': lambda x: x,
+    }
+
+    with run(
+        [
+            'git', 'annex', 'find',
+            # we want everything regardless of local availability.
+            # the faster `--anything` came only in 10.20230126
```

Comment:
I am not sure how much of a performance difference it makes, but since we know the annex version we could choose the faster command on applicable annex versions.